spec/02-integration/14-tracing/02-propagation_spec.lua

```diff
@@ -2,165 +2,167 @@ local helpers = require "spec.helpers"
+          "plugins",
+        }))
+
+        local multi_plugin_route = bp.routes:insert({
+          hosts = { "propagate.test" },
         })
-        proxy_client = helpers.proxy_client()
-      end)
-
-      lazy_teardown(function()
-        if proxy_client then
-          proxy_client:close()
-        end
-        helpers.stop_kong()
-      end)
 
-      it("propagates the balancer span", function ()
-        local thread = helpers.tcp_server(TCP_PORT)
-        local r = assert(proxy_client:send {
-          method  = "GET",
-          path = "/request",
-          headers = {
-            ["Host"] = "propagate.test",
-          }
+        bp.plugins:insert({
+          name = "pre-function",
+          route = multi_plugin_route,
+          config = generate_function_plugin_config({
+            extract = "{}",               -- ignores incoming
+            inject = '{ "preserve" }',    -- falls back to default
+            default_format = '"b3-single"',      -- defaults to b3
+          }, trace_id, span_id),
         })
-        assert.res_status(200, r)
-        local body = r:read_body()
-        body = assert(body and cjson.decode(body))
-
-        local ok, res = thread:join()
-        assert.True(ok)
-        assert.is_string(res)
-
-        -- expected spans are returned
-        local spans = cjson.decode(res)
-        assert.is_same(2, #spans, res)
-        local balancer_span = spans[2]
-        assert.is_same("kong.balancer", balancer_span.name)
-
-        local traceparent = assert(body.headers.traceparent)
-        local trace_id = balancer_span.trace_id
-        local span_id = balancer_span.span_id
-        -- traceparent contains correct trace id and the balancer span's id
-        assert.equals("00-" .. trace_id .. "-" .. span_id .. "-01", traceparent)
-      end)
-    end)
 
-    describe("parsing incoming headers", function ()
-      local trace_id = gen_id(16)
-      local span_id = gen_id(8)
+        bp.plugins:insert({
+          name = "post-function",
+          route = multi_plugin_route,
+          config = generate_function_plugin_config({
+            extract = '{ "w3c", "b3" }',      -- reads b3
+            inject = '{ "w3c" }',             -- and injects w3c
+            default_format = "datadog",       -- default not used here
+            clear = '{ "ot-tracer-spanid" }',     -- clears this header
+          }),
+        })
 
-      lazy_setup(function()
-        assert(helpers.start_kong {
+        helpers.start_kong({
           database = strategy,
+          plugins = "bundled",
           nginx_conf = "spec/fixtures/custom_nginx.template",
-          plugins = "tcp-trace-exporter,trace-propagator",
-          tracing_instrumentations = "request,router,balancer,plugin_access,plugin_header_filter",
-          tracing_sampling_rate = 1,
+          untrusted_lua = "on",
         })
         proxy_client = helpers.proxy_client()
       end)
 
-      lazy_teardown(function()
+      teardown(function()
```

Comment:
Why using `teardown` instead of `lazy_teardown`?

Reply:
fixed